domain/src/Aggregates/SharePoolingAsset/Entities/SharePoolingAssetAcquisition.php

```diff
@@ -63,8 +63,23 @@ public function section104PoolCostBasis(): FiatAmount
         return $this->costBasis->dividedBy($this->quantity)->multipliedBy($this->section104PoolQuantity());
     }
 
-    /** Increase the same-day quantity and adjust the 30-day quantity accordingly. */
     public function increaseSameDayQuantity(Quantity $quantity): self
+    {
+        if ($quantity->isGreaterThan($availableQuantity = $this->section104PoolQuantity())) {
+            throw SharePoolingAssetAcquisitionException::insufficientSameDayQuantityToIncrease($quantity, $availableQuantity);
+        }
+
+        $this->sameDayQuantity = $this->sameDayQuantity->plus($quantity);
+
+        return $this;
+    }
+
+    /**
+     * Increase the same-day quantity and adjust the 30-day quantity accordingly.
+     *
+     * @return Quantity the added quantity
+     */
+    public function increaseSameDayQuantityUpToAvailableQuantity(Quantity $quantity): Quantity
```

Comment:
ℹ️ There are now two separate methods – `increaseSameDayQuantity` and `increaseSameDayQuantityUpToAvailableQuantity`.

The former will throw an exception if the specified quantity exceeds the currently unallocated same-day quantity and is used by the `QuantityAdjuster` service (see further down).

The latter will allocate as much as possible the specified quantity, also reallocating some or all of the 30-day quantity to the same-day quantity as necessary.